modules/tensor_mechanics/src/constraints/NodalFrictionalConstraint.C

```diff
@@ -0,0 +1,342 @@
+                                                  mesh_inserter_iterator<Elem>(_mesh.getMesh()));
+
+    _mesh.update(); // Rebuild node_to_elem_map
+
+    // Find elems again now that we know they're there
+    const auto & new_node_to_elem_map = _mesh.nodeToElemMap();
+    auto node_to_elem_pair = new_node_to_elem_map.find(_master_node_vector[0]);
+    bool found_elems = (node_to_elem_pair != new_node_to_elem_map.end());
+
+    if (!found_elems)
+      mooseError("Colundn't find any elements connected to master node.");
+
+    for (unsigned int i = 0; i < _master_node_vector.size(); ++i)
+      elems[i] = node_to_elem_pair->second;
+  }
+  else // serial mesh
+  {
+    for (unsigned int i = 0; i < _master_node_vector.size(); ++i)
+    {
+      auto node_to_elem_pair = node_to_elem_map.find(_master_node_vector[i]);
+      bool found_elems = (node_to_elem_pair != node_to_elem_map.end());
+
+      if (!found_elems)
+        mooseError("Couldn't find any elements connected to master node");
+
+      elems[i] = node_to_elem_pair->second;
+    }
+  }
+
+  for (unsigned int i = 0; i < _master_node_vector.size(); ++i)
+  {
+    if (elems[i].size() == 0)
+      mooseError("Couldn't find any elements connected to master node");
+
+    for (unsigned int j = 0; j < elems[i].size(); ++j)
+      _subproblem.addGhostedElem(elems[i][j]);
+  }
+
+  // Cache map between slave node and master node
+  _connected_nodes.clear();
+  _master_conn.clear();
+  for (unsigned int j = 0; j < slave_nodelist.size(); ++j)
+  {
+    if (_mesh.nodeRef(slave_nodelist[j]).processor_id() == _subproblem.processor_id())
+    {
+      Node & slave_node = _mesh.nodeRef(slave_nodelist[j]);
+      for (unsigned int i = 0; i < _master_node_vector.size(); ++i)
+      {
+        Node & master_node = _mesh.nodeRef(_master_node_vector[i]);
+        Real d = (slave_node - master_node).norm();
+        if (MooseUtils::absoluteFuzzyEqual(d, 0.0))
+        {
+          _master_conn.push_back(i);
+          _connected_nodes.push_back(slave_nodelist[j]);
+          break;
+        }
+      }
+    }
+    /*if (_master_conn[j] == std::numeric_limits<unsigned int>::max())
+    {
+        printf("in here \n");
+        mooseError("No master node located at the same elevation as the slave node.");
+    }*/
+  }
+  printf("total slave nodes, master nodes: %lu, %lu \n",
+         _master_conn.size(),
+         _master_node_vector.size());
+}
+
+void
+NodalFrictionalConstraint::computeResidual(NumericVector<Number> &
+                                           /*residual*/)
+{
+  std::vector<dof_id_type> masterdof = _var.dofIndices();
+  std::vector<dof_id_type> slavedof = _var.dofIndicesNeighbor();
+  DenseVector<Number> re(masterdof.size());
+  DenseVector<Number> neighbor_re(slavedof.size());
+
+  re.zero();
+  neighbor_re.zero();
+
+  for (_i = 0; _i < slavedof.size(); ++_i)
+  {
+    _j = _master_conn[_i];
+    switch (_formulation)
```

Comment:
I'd drop this check for the formulation here.  In the constructor, you are making sure that they specified the only type that works currently.

Reply:
Done.